File: app/src/main/java/com/sasfmlzr/findfilm/fragment/discoverfilm/DiscoverFilmFragment.java
package com.sasfmlzr.findfilm.fragment.discoverfilm;

import android.content.Context;
import android.database.MatrixCursor;
import android.databinding.BindingAdapter;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.TextView;

import com.sasfmlzr.findfilm.R;
import com.sasfmlzr.findfilm.adapter.SearchAdapter;
import com.sasfmlzr.findfilm.adapter.VerticalItemDecoration;
import com.sasfmlzr.findfilm.databinding.DiscoverFragmentBinding;
import com.sasfmlzr.findfilm.request.DiscoverMovieRequest;

import java.util.List;
import java.util.Objects;
import java.util.Timer;
import java.util.TimerTask;

public class DiscoverFilmFragment extends Fragment {
    private static final String QUERY_SEARCH_ARGS = "querySearch";
    private String querySearch;

    private DiscoverFilmViewModel viewModel;
    private DiscoverFragmentBinding viewDataBinding;
    public DiscoverFilmFragment.OnFilmSelectedListener filmSelectedListener;


    public View view;
    public SearchView searchView;
    public Bundle savedState = null;
    private Timer timer;
    private Menu menu;

    public static DiscoverFilmFragment newInstance(String querySearch) {
        Bundle args = new Bundle();
        args.putString(QUERY_SEARCH_ARGS, querySearch);
        DiscoverFilmFragment fragment = new DiscoverFilmFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            querySearch = getArguments().getString(QUERY_SEARCH_ARGS);
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.discover_fragment, container, false);
        setHasOptionsMenu(true);

        savedState = null;

        viewModel = new DiscoverFilmViewModel(filmSelectedListener);
        viewDataBinding = DiscoverFragmentBinding.bind(view);
        viewDataBinding.setViewmodel(viewModel);
        loadRecyclerFilmView();

        viewModel.runRequest(querySearch);

        return viewDataBinding.getRoot();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        savedState = saveState();
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        if (searchView != null) {
            outState.putString("currentSearchQuery", searchView.getQuery().toString());
        }
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        try {
            filmSelectedListener = (DiscoverFilmFragment.OnFilmSelectedListener) getParentFragment();
        } catch (ClassCastException e) {
            throw new ClassCastException(context.toString() + " OnFilmSelectedListener not attached");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        filmSelectedListener = null;
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        this.menu = menu;

        searchView = (SearchView) menu.findItem(R.id.search).getActionView();
        searchView.setOnQueryTextListener(new android.support.v7.widget.SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                if (query != null) {
                    filmSelectedListener.filmSearched(query);
                    InputMethodManager imm = (InputMethodManager)
                            Objects.requireNonNull(getActivity())
                                    .getSystemService(Context.INPUT_METHOD_SERVICE);
                    assert imm != null;
                    imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
                }
                return true;
            }

            @Override
            public boolean onQueryTextChange(String query) {
                loadHistory(query);
                return true;
            }
        });
        super.onCreateOptionsMenu(menu, inflater);
    }

    public interface RecyclerElementEnded {
        void isEnded();
    }

    public interface SearchCallback {
        void isFind(List<DiscoverMovieRequest.Result> filmList);
    }

    public interface OnFilmSelectedListener {
        void filmClicked(int idFilm);

        void filmSearched(String query);
    }

    public interface FilmListComplete {
        void isCompleted(List<DiscoverMovieRequest.Result> filmList);
    }

    public void loadRecyclerFilmView() {
        viewDataBinding.discoverFilmList.setLayoutManager(new LinearLayoutManager(view.getContext()));
        viewDataBinding.discoverFilmList.addItemDecoration(new VerticalItemDecoration(50));
    }

    public Bundle saveState() {
        Bundle state = new Bundle();
        if (searchView != null) {
            state.putString("currentSearchQuery", searchView.getQuery().toString());
        }
        return state;
    }

    private void loadHistory(String query) {
        String[] columns = new String[]{"_id", "text"};
        Object[] temp = new Object[]{0, "default"};
        final SearchView search = (SearchView) menu.findItem(R.id.search).getActionView();
        MatrixCursor cursor = new MatrixCursor(columns);
        if (query.length() == 0) {
            search.setSuggestionsAdapter(new SearchAdapter(getContext(),
                    cursor,
                    null,
                    filmSelectedListener));
            return;
        }
        if (query.length() > 2) {
            SearchCallback callback = filmList -> {
                for (int i = 0; i < filmList.size(); i++) {
                    temp[0] = i;
                    temp[1] = filmList.get(i);
                    cursor.addRow(temp);
                }
                search.setSuggestionsAdapter(new SearchAdapter(getContext(),
                        cursor,
                        filmList,
                        filmSelectedListener));
            };
            Handler handler = new Handler();
            TimerTask timerTask = new TimerTask() {
                @Override
                public void run() {
                    handler.post(() -> {
                        viewModel.runSearchRequestFilm(query, callback);
                        timer = null;
                    });
                }
            };
            if (timer != null) {
                timer.cancel();
            }
            timer = new Timer();
            timer.schedule(timerTask, 2000);
        }
    }
}

File: app/src/main/java/com/sasfmlzr/findfilm/fragment/discoverfilm/DiscoverFilmFragment.kt
package com.sasfmlzr.findfilm.fragment.discoverfilm

import android.content.Context
import android.database.MatrixCursor
import android.os.Bundle
import android.os.Handler
import android.view.*
import android.view.inputmethod.InputMethodManager
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import com.sasfmlzr.findfilm.R
import com.sasfmlzr.findfilm.adapter.SearchAdapter
import com.sasfmlzr.findfilm.adapter.VerticalItemDecoration
import com.sasfmlzr.findfilm.databinding.DiscoverFragmentBinding
import com.sasfmlzr.findfilm.request.DiscoverMovieRequest
import java.util.*

class DiscoverFilmFragment : Fragment() {
    private var querySearch: String? = null
    private var viewModel: DiscoverFilmViewModel? = null
    private var viewDataBinding: DiscoverFragmentBinding? = null
    var filmSelectedListener: OnFilmSelectedListener? = null
    var searchView: SearchView? = null
    var savedState: Bundle? = null
    private var timer: Timer? = null
    private var menu: Menu? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            querySearch = requireArguments().getString(QUERY_SEARCH_ARGS)
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        viewDataBinding = DiscoverFragmentBinding.bind(inflater.inflate(R.layout.discover_fragment, container, false))
        setHasOptionsMenu(true)
        savedState = null
        viewModel = DiscoverFilmViewModel(filmSelectedListener)
       viewDataBinding!!.viewmodel = viewModel
        loadRecyclerFilmView()
        viewModel!!.runRequest(querySearch)
        return viewDataBinding!!.root
    }

    override fun onDestroyView() {
        super.onDestroyView()

        savedState = saveState()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        if (searchView != null) {
            outState.putString("currentSearchQuery", searchView!!.query.toString())
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        filmSelectedListener = try {
            parentFragment as OnFilmSelectedListener?
        } catch (e: ClassCastException) {
            throw ClassCastException("$context OnFilmSelectedListener not attached")
        }
    }

    override fun onDetach() {
        super.onDetach()
        filmSelectedListener = null
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        this.menu = menu
        searchView = menu.findItem(R.id.search).actionView as SearchView
        searchView!!.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                filmSelectedListener!!.filmSearched(query)
                val imm = requireActivity().getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                imm.hideSoftInputFromWindow(view!!.windowToken, 0)
                return true
            }

            override fun onQueryTextChange(query: String): Boolean {
                loadHistory(query)
                return true
            }
        })
        super.onCreateOptionsMenu(menu, inflater)
    }

    interface RecyclerElementEnded {
        val isEnded: Unit
    }

    interface SearchCallback {
        fun isFind(filmList: List<DiscoverMovieRequest.Result>)
    }

    interface OnFilmSelectedListener {
        fun filmClicked(idFilm: Int)
        fun filmSearched(query: String?)
    }

    interface FilmListComplete {
        fun isCompleted(filmList: List<DiscoverMovieRequest.Result>)
    }

    private fun loadRecyclerFilmView() {
        viewDataBinding!!.discoverFilmList.layoutManager = LinearLayoutManager(requireContext())
        viewDataBinding!!.discoverFilmList.addItemDecoration(VerticalItemDecoration(50))
    }

    private fun saveState(): Bundle {
        val state = Bundle()
        if (searchView != null) {
            state.putString("currentSearchQuery", searchView!!.query.toString())
        }
        return state
    }

    private fun loadHistory(query: String) {
        val columns = arrayOf("_id", "text")
        val temp = arrayOf<Any>(0, "default")
        val search = menu!!.findItem(R.id.search).actionView as SearchView
        val cursor = MatrixCursor(columns)
        if (query.isEmpty()) {
            search.suggestionsAdapter = SearchAdapter(
                context,
                cursor,
                null,
                filmSelectedListener!!
            )
            return
        }
        if (query.length > 2) {
            val callback = object: SearchCallback{
                override fun isFind(filmList: List<DiscoverMovieRequest.Result>) {
                    for (i in filmList.indices) {
                        temp[0] = i
                        temp[1] = filmList[i]
                        cursor.addRow(temp)
                    }
                    search.suggestionsAdapter = SearchAdapter(
                        context,
                        cursor,
                        filmList,
                        filmSelectedListener!!
                    )
                }
            }

            val handler = Handler()
            val timerTask: TimerTask = object : TimerTask() {
                override fun run() {
                    handler.post {
                        viewModel!!.runSearchRequestFilm(query, callback)
                        timer = null
                    }
                }
            }
            if (timer != null) {
                timer!!.cancel()
            }
            timer = Timer()
            timer!!.schedule(timerTask, 2000)
        }
    }

    companion object {
        private const val QUERY_SEARCH_ARGS = "querySearch"
        fun newInstance(querySearch: String?): DiscoverFilmFragment {
            val args = Bundle()
            args.putString(QUERY_SEARCH_ARGS, querySearch)
            val fragment = DiscoverFilmFragment()
            fragment.arguments = args
            return fragment
        }
    }
}